pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,541 @@
+		_,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		Twox64Concat,
+		T::AssetCurrencyId,
+		BoundedVec<T::Hash, ConstU32<1_000_000>>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		OrderCreated {
+			order_id: T::Hash,
+			creator_account: T::AccountId,
+			currency_in: T::AssetCurrencyId,
+			currency_out: T::AssetCurrencyId,
+			buy_amount: T::ForeignCurrencyBalance,
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+			sell_price_limit: T::ForeignCurrencyBalance,
+		},
+		OrderCancelled {
+			account: T::AccountId,
+			order_id: T::Hash,
+		},
+		OrderUpdated {
+			order_id: T::Hash,
+			account: T::AccountId,
+			buy_amount: T::ForeignCurrencyBalance,
+			sell_price_limit: T::ForeignCurrencyBalance,
+			min_fullfillment_amount: T::ForeignCurrencyBalance,
+		},
+		OrderFulfillment {
+			order_id: T::Hash,
+			placing_account: T::AccountId,
+			fulfilling_account: T::AccountId,
+			partial_fulfillment: bool,
+			fulfillment_amount: T::ForeignCurrencyBalance,
+			currency_in: T::AssetCurrencyId,
+			currency_out: T::AssetCurrencyId,
+			sell_price_limit: T::ForeignCurrencyBalance,
+		},
+	}
+
+	#[pallet::error]
+	#[derive(PartialEq)]
+	pub enum Error<T> {
+		AssetPairOrdersOverflow,
+		ConflictingAssetIds,
+		InsufficientAssetFunds,
+		InsufficientReserveFunds,
+		InvalidBuyAmount,
+		InvalidMinPrice,
+		InvalidAssetId,
+		OrderNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<T as frame_system::Config>::Hash: PartialEq<<T as frame_system::Config>::Hash>,
+	{
+		/// Create an order, with the minimum fulfillment amount set to the buy
+		/// amount, as the first iteration will not have partial fulfillment
+		#[pallet::call_index(0)]
+		// dummy weight for now
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn create_order_v1(
+			origin: OriginFor<T>,
+			asset_in: T::AssetCurrencyId,
+			asset_out: T::AssetCurrencyId,
+			buy_amount: T::ForeignCurrencyBalance,
+			price: T::ForeignCurrencyBalance,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			Self::place_order(
+				account_id, asset_in, asset_out, buy_amount, price, buy_amount,
+			)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		// dummy weight for now
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn user_cancel_order(origin: OriginFor<T>, order_id: T::Hash) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			// verify order matches account
+			// UserOrders using Resultquery, if signed account
+			// does not match user for order id, we will get an Err Result
+			Self::cancel_order(order_id)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(2)]
+		// dummy weight for now
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn fill_order_full(origin: OriginFor<T>, order_id: T::Hash) -> DispatchResult {
```

Comment:
Are there plans to partially fill orders? I mean assume one big ask order comes in at the best price (lowest ask). Now a small fish can't really buy at this price, because the order is too big for the small fish.

The other case is the trickier: Assume there are many small ask orders at the best price (lowest ask). Now a big whale places a market order and buys all small ask orders. First, this often creates slippage, as you might know already.. Second, there might be not enough open ask orders to fill the big bid of the whale. Third, the whale needs to transact with many small fish accounts to fulfil the small orders. This leads to a high computational demand.

So is your plan to just allow the orders to be filled 100% to mitigate this and sacrifice that a small fish can fill the possible big order? 

There are two